plan/planbuilder.go

```diff
@@ -951,154 +935,251 @@ func (b *planBuilder) buildInsert(insert *ast.InsertStmt) Plan {
+		return
+	}
+	for _, tCol := range tCols {
+		if tCol.IsGenerated() {
+			b.err = ErrBadGeneratedColumn.GenByArgs(tCol.Name.O, tableInfo.Name.O)
+			return
 		}
+	}
 
+	for i, assign := range insert.Setlist {
 		expr, _, err := b.rewriteWithPreprocess(assign.Expr, mockTablePlan, nil, true, checkRefColumn)
 		if err != nil {
 			b.err = errors.Trace(err)
-			return nil
 		}
 		insertPlan.SetList = append(insertPlan.SetList, &expression.Assignment{
-			Col:  col,
+			Col:  exprCols[i],
 			Expr: expr,
 		})
 	}
-
 	insertPlan.Schema4OnDuplicate = insertPlan.tableSchema
-	if insert.Select != nil {
-		selectPlan := b.build(insert.Select)
-		if b.err != nil {
-			return nil
-		}
+}
+
+func (b *planBuilder) buildValuesListOfInsert(insert *ast.InsertStmt, insertPlan *Insert, mockTablePlan *LogicalTableDual,
+	checkRefColumn func(n ast.Node) ast.Node) {
+	affectedValuesCols, err := insertPlan.getAffectCols(insert)
+	if err != nil {
+		b.err = errors.Trace(err)
+		return
+	}
 
-		numInsertCols := mathutil.Min(selectPlan.Schema().Len(), len(tableInfo.Columns))
-		// If the column to be inserted in the insert table is a generated
-		// column, raises a "ErrBadGeneratedColumn" error here.
-		for _, col := range tableInfo.Columns[:numInsertCols] {
+	if len(insert.Columns) > 0 || len(insert.Lists[0]) > 0 {
+		// If value_list is not empty or the col_list is not empty, length of value_list should be the same with col_list's.
+		if len(insert.Lists[0]) != len(affectedValuesCols) {
+			b.err = ErrWrongValueCountOnRow.GenByArgs(1)
+			return
+		}
+		// No generated column is allowed.
+		for _, col := range affectedValuesCols {
 			if col.IsGenerated() {
-				b.err = ErrBadGeneratedColumn.GenByArgs(col.Name.O, tableInfo.Name.O)
-				return nil
+				b.err = ErrBadGeneratedColumn.GenByArgs(col.Name.O, insertPlan.Table.Meta().Name.O)
+				return
 			}
 		}
+		// If the value_list and col_list is empty and we have generated column, we can still write to this table.
+		// i.e. insert into t values(); can be executed successfully if t have generated column.
+	}
 
-		insertPlan.SelectPlan, b.err = doOptimize(b.optFlag, selectPlan.(LogicalPlan))
-		if b.err != nil {
-			return nil
+	totalTableCols := insertPlan.Table.Cols()
+	for i, valuesItem := range insert.Lists {
+		// The length of the value_list should keep unchanged.
+		// "insert into t values (), ()" is valid.
+		// "insert into t values (), (1)" is not valid.
+		// "insert into t values (1), ()" is not valid.
+		// "insert into t values (1,2), (1)" is not valid.
+		if i > 0 && len(insert.Lists[i-1]) != len(valuesItem) {
+			b.err = ErrWrongValueCountOnRow.GenByArgs(i + 1)
+			return
 		}
-
-		insertPlan.Schema4OnDuplicate = expression.MergeSchema(insertPlan.tableSchema, insertPlan.SelectPlan.Schema())
+		exprList := make([]expression.Expression, 0, len(valuesItem))
+		for j, valueItem := range valuesItem {
+			var expr expression.Expression
+			var err error
+			if dft, ok := valueItem.(*ast.DefaultExpr); ok {
```

Comment:
Usually the abbreviation for `default` is `def`, see https://www.abbreviations.com/abbreviation/Default